rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,382 @@
+
+### Filesystem
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  ├── etc
+   │  │  └── systemd
+   │  │     └── teleport.service
+   │  └── backup
+   │     ├── teleport
+   │     └── backup.yaml
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/tsh
+/usr/local/bin/tsh -> /var/lib/teleport/versions/15.0.0/bin/tsh
+$ ls -l /usr/local/bin/tbot
+/usr/local/bin/tbot -> /var/lib/teleport/versions/15.0.0/bin/tbot
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport-updater
+/usr/local/bin/teleport-updater -> /var/lib/teleport/versions/15.0.0/bin/teleport-updater
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+kind: agent_versions
+spec:
+  # proxy specifies the Teleport proxy address to retrieve the agent version and update configuration from.
+  proxy: mytenant.teleport.sh
+  # enabled specifies whether auto-updates are enabled, i.e., whether teleport-updater update is allowed to update the agent.
+  enabled: true
+  # active_version specifies the active (symlinked) deployment of the telepport agent.
+  active_version: 15.1.1
+```
+
+backup.yaml:
+```
+version: v1
+kind: config_backup
+spec:
+  # proxy address from the backup
+  proxy: mytenant.teleport.sh
+  # version from the backup
+  version: 15.1.0
+  # time the backup was created
+  creation_time: 2020-12-09T16:09:53+00:00
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-updater update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-updater` command:
+```shell
+$ teleport-updater enable --proxy mytenant.teleport.sh
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used.
+
+On servers without Teleport installed already, the `enable` subcommand will change the behavior of `teleport-update update` to update teleport and restart the existing agent, if running.
+It will also run update teleport immediately, to ensure that subsequent executions succeed.
+
+The `enable` subcommand will:
+1. Configure `updates.yaml` with the current proxy address and set `enabled` to true.
```

Comment:
Hitting the ping endpoint first would be safer as we can catch connectivity issues here, exit with a nice error message, and not change the config.

Reply:
Moved this to very end, so an upgrade has to succeed before it's enabled.